### ISSUE TITLE
txg_sync and z_rd_int/* become top I/O consumers, the host renders unusable for a period of time

### ISSUE THREAD

**Issue by alexclear** [zfsonlinux/zfs#1538]
I have a dual Xeon E5645 system w/8 SATA disks, /boot is on an USB drive and / is on ZFS:

```
root@debhost1 ~ zfs list
NAME                                 USED  AVAIL  REFER  MOUNTPOINT
vol0                                2.50T  1.96T   350M  /vol0
vol0/kvm-Common-Host                 775G  1.96T   186G  /var/lib/libvirt/images/Common-Host
vol0/kvm-Dev6-Host                  9.28G  1.96T  9.28G  /var/lib/libvirt/images/Dev6-Host
vol0/kvm-Dev7-Host                  9.25G  1.96T  9.25G  /var/lib/libvirt/images/Dev7-Host
vol0/kvm-HBase-Slave-1              99.1G  1.96T  23.1G  /var/lib/libvirt/images/HBase-Slave-1
vol0/kvm-HBase-Slave-2               107G  1.96T  23.8G  /var/lib/libvirt/images/HBase-Slave-2
vol0/kvm-W2K8                        577G  1.96T   184G  /var/lib/libvirt/images/W2K8
vol0/libvirt-images                  130G  1.96T    31K  /vol0/libvirt-images
vol0/libvirt-images/Alex-Client     23.2G  1.96T  9.74G  /var/lib/libvirt/images/Alex-Client
vol0/libvirt-images/Dev3-Host       27.1G  1.96T  27.1G  /var/lib/libvirt/images/Dev3-Host
vol0/libvirt-images/Dev4-Host       23.6G  1.96T  23.6G  /var/lib/libvirt/images/Dev4-Host
vol0/libvirt-images/Dev5-Host       20.2G  1.96T  20.2G  /var/lib/libvirt/images/Dev5-Host
vol0/libvirt-images/Dev7-Host       91.2M  1.96T  91.2M  /vol0/libvirt-images/Dev7-Host
vol0/libvirt-images/Dev8-Host       2.73G  1.96T  2.73G  /var/lib/libvirt/images/Dev8-Host
vol0/libvirt-images/HBase-Master    9.65G  1.96T  5.23G  /var/lib/libvirt/images/HBase-Master
vol0/libvirt-images/W7-01d          9.97G  1.96T  9.97G  /var/lib/libvirt/images/W7-01d
vol0/libvirt-images/W7-Scanner-New  12.9G  1.96T  12.9G  /var/lib/libvirt/images/W7-Scanner-New
vol0/rootfs                          260G  1.96T   189G  /
vol0/var                             579G  1.96T   579G  /var2
root@debhost1 ~ 
```

The zpool is:

```
root@debhost1 ~ zpool status
  pool: vol0
 state: ONLINE
  scan: scrub repaired 0 in 44h43m with 0 errors on Mon May 13 00:06:19 2013
config:

    NAME                                           STATE     READ WRITE CKSUM
    vol0                                           ONLINE       0     0     0
      mirror-0                                     ONLINE       0     0     0
        scsi-SATA_WDC_WD1003FBYX-_WD-WMAW30418757  ONLINE       0     0     0
        scsi-SATA_WDC_WD1003FBYX-_WD-WMAW31056873  ONLINE       0     0     0
      mirror-1                                     ONLINE       0     0     0
        scsi-SATA_WDC_WD1003FBYX-_WD-WMAW31066713  ONLINE       0     0     0
        scsi-SATA_WDC_WD1003FBYX-_WD-WCAW33977049  ONLINE       0     0     0
      mirror-2                                     ONLINE       0     0     0
        scsi-SATA_WDC_WD1003FBYX-_WD-WMAW31077083  ONLINE       0     0     0
        scsi-SATA_WDC_WD1003FBYX-_WD-WCAW35326792  ONLINE       0     0     0
      mirror-3                                     ONLINE       0     0     0
        scsi-SATA_WDC_WD2000FYYZ-_WD-WCC1P0020205  ONLINE       0     0     0
        scsi-SATA_WDC_WD2000FYYZ-_WD-WCC1P0021665  ONLINE       0     0     0
    cache
      ata-OCZ-AGILITY4_7M7U6Y6X7E6Q3K2D3E3U-part1  ONLINE       0     0     0

errors: No known data errors
root@debhost1 ~ 
```

This is a Debian 7.0-based system w/ZoL 0.6.1-1:

```
root@debhost1 ~ uname -a
Linux debhost1 3.2.0-4-amd64 #1 SMP Debian 3.2.41-2 x86_64 GNU/Linux
```

It is used mainly for virtualization, most datasets have sync=disabled and recordsize=4K
ARC is capped by 20G, ARC utilization graph follows:

![zfs_stats_utilization-day](https://f.cloud.github.com/assets/857605/680722/94abdae2-d999-11e2-9c29-1adfc691e2eb.png)

L2ARC size and efficiency graphs follow:

![zfs_stats_l2utilization-day](https://f.cloud.github.com/assets/857605/680725/be14b048-d999-11e2-8e07-e5a3dcf7bede.png)
![zfs_stats_l2efficiency-day](https://f.cloud.github.com/assets/857605/680727/c2559dc0-d999-11e2-9c27-a4544ddbe5e5.png)

Disk load is almost constant during working hours but we suffer from sporadical performance drops which do not seem to correlate to actual load generated by virtual machines. Or at least we were unable to find the correlation yet.
Output of iotop -d 5 -P looks like this during these performance drop periods:

![screenshot-txg_sync](https://f.cloud.github.com/assets/857605/680752/9b83ff6a-d99a-11e2-8b8a-d72b02d6c12c.png)

We also collected iostat -x -k 2 20 output here: https://gist.github.com/alexclear/5821955 and zpool iostat -v 2 20 (at the same time) output here: https://gist.github.com/alexclear/5821960

It seems like the system experiences a lot of internal I/O (mostly random reads) not consumed directly by userspace processes during these periods. It is also very odd that L2ARC efficiency remains quite low.

**Comment by pdf** [comment]
I appear to be seeing this on 0.6.5.3 - load average is currently 40+, pool is doing 1GB/s read throughput, all of it generated by `z_rd_*`.

If someone notes this in the very near future and there are some details I can capture while this is occuring, please let me know.  Otherwise, I'm going to have to bounce this rather soon, `iotop -P`:

![iotop](https://cloud.githubusercontent.com/assets/146192/11160427/b9b3efdc-8abd-11e5-9fce-20da51b27a8c.png)

**Comment by pdf** [comment]
Whoops, just realized it's scrubbing.  Still, it seems like the scrub is starving other processes  (load avg 40, 35-40% iowait), which I don't recall being this dramatic a problem in the past.

**Comment by behlendorf** [comment]
@pdf this may be a result of the change in scheduling priority for the ZFS threads, 1229323d5f82bcb0525b312c80330968cd41466e and an increase in their number aa9af22cdf8d16c197974c3a478d2053b3bed498.

**Comment by pdf** [comment]
Seem to be likely suspects.  This does make scrubs impractical in production, which is problematic.

**Comment by behlendorf** [comment]
Some tuning may be required, but real user IO will still be given priority.

**Comment by aktau** [comment]
Oddly, I've also noticed something similar, but I verified that this was not happening during a scrub:

```bash
sudo pidstat -d 5
Linux 4.2.0-1-amd64 (kdi) 	12/06/2015 	_x86_64_	(2 CPU)

12:16:34 AM   UID       PID   kB_rd/s   kB_wr/s kB_ccwr/s iodelay  Command
12:16:39 AM     0        31      0.00      0.00      0.00       1  kswapd0
12:16:39 AM     0       150      0.00     16.00      0.00       1  jbd2/sda1-8
12:16:39 AM     0       669   1838.40      0.00      0.00       0  z_rd_int_0
12:16:39 AM     0       670   2048.00      0.00      0.00       0  z_rd_int_1
12:16:39 AM     0       671   1971.20      0.00      0.00       0  z_rd_int_2
12:16:39 AM     0       672   2037.60      0.00      0.00       0  z_rd_int_3
12:16:39 AM     0       673   1715.20      0.00      0.00       0  z_rd_int_4
12:16:39 AM     0       674   2068.80      0.00      0.00       0  z_rd_int_5
12:16:39 AM     0       675   1664.00      0.00      0.00       0  z_rd_int_6
12:16:39 AM     0       676   1995.20      0.00      0.00       0  z_rd_int_7
12:16:39 AM     0       714      0.00      0.00      0.00     245  txg_sync
12:16:39 AM   111       894   2236.80    931.20      0.00       0  transmission-da
12:16:39 AM  1000       931     36.00      0.00      0.00     158  kodi.bin
```

I'm on 0.6.5.2 with spl 0.6.5.1 (debian 9), should that be important.

I actually suspected that the "reads" from transmission-da were somehow being multiplexed to the z_rd_int_* kernel threads. But the numbers don't add up.

Note that my ZFS pool consists of just one single USB hard drive. No raidz, no mirroring, nothing fancy. So parity data calculation or others things (should the rd_int threads do that) can't be it.

**Comment by aktau** [comment]
After reading #2952, I feel obliged to say that I'm using lz4 compression on my zpool. Here's the configuration: https://gist.github.com/aktau/fa6b5ef2fd40201a489b

**Comment by behlendorf** [comment]
@aktau you've hit #3904 and you seem to have built with --enable-debug.  You can safely avoid your issue for the moment by not enabling debug, see commit 935434e.

**Comment by aktau** [comment]
> @aktau you've hit #3904 and you seem to have built with --enable-debug. You can safely avoid your issue for the moment by not enabling debug, see commit 935434e.

Thanks for the update! I'm not running a self-built kernel module though, I just grabbed it from the Debian Jessie repo the zfsonlinux project provides. I assume those are built with debugging support enabled, then?

That means that the `z_rd_int` high IO load is a separate thing, sorry for the confusion. Though, from reading #3904, the two seem related: high read workload with a few random writes cause it to happen more often, says a poster. Might be because my unit has low memory and the ARC is filling up available memory quite fast because of all the reading.

That said, I find it odd that my particular workload is causing this. Back-of-the-envelope, the only process that is doing significant reading is transmission-da. Apart from checksum verification and other assorted things it's doing, the main source of reads is sharing chunks. I've capped that to 100KB/sec (so give or a take a few 10's of KB/sec). The HDD, slow and old though it is, should definitely be able to sustain that without reaching 50-90% utilization, I think.

**Comment by behlendorf** [comment]
Is anybody still observing this behavior with a recent version of ZFS?

**Comment by hooliowobbits** [comment]
Hi yes, i installed ZOL today on Debian Jessie, along with proxmox VE 4.3-3/557191d3.  

i am currently doing an debian install within a vm and the copy was slowing down and down and started wondering about host performance.  i'm then ran `$ dd if=/dev/zero of=/tank/perftest/dd bs=1024M count=1 oflag=dsync`. now i/o seems to have dried up altogether :(

syslog showing lots and lots of `Oct 18 22:03:30 riverbed kernel: [ 3480.297060] INFO: task txg_sync:6052 blocked for more than 120 seconds.
Oct 18 22:03:30 riverbed kernel: [ 3480.297410] txg_sync        D ffff8801fbe1faa8     0  6052      2 0x00000000
Oct 18 22:03:30 riverbed kernel: [ 3480.297749]  [<ffffffffc049aa74>] txg_sync_thread+0x3e4/0x6a0 [zfs]
Oct 18 22:03:30 riverbed kernel: [ 3480.297811]  [<ffffffffc049a690>] ? txg_sync_stop+0xf0/0xf0 [zfs]`

**Comment by behlendorf** [comment]
@hooliowobbits can you post your ZFS version.  You can find it in `/sys/module/zfs/version`.

**Comment by hooliowobbits** [comment]
0.6.5.7-10_g27f3ec9

**Comment by kpande** [comment]
@hooliowobbits what is your ashift value?

**Comment by hooliowobbits** [comment]
`$ sudo zpool get all tank | grep ashift
tank  ashift                      0                           default`

admittedly i had seen other users altering ashift, but not knowing much about that i left it as default

**Comment by kettenbach-it** [comment]
I think I have the same problem.

The system is a Hexa-Core-Xeon E5-1650 with 128GB RAM and 2x 2TB SATA Disk running Kernelversion 4.4.40-1-pve and Proxmox 4.4-12 with ZFS module version 0.6.5.9-1.
Compression is on. Dedup is of and ashift=12. The SATA disks are in a mirror running the system and some volumes. The systems has an additional NVM which is supposed to work as Log+Cache, but it's not enabled right now and it doesn't cause the problem.

From time to time, tenth of z_wr_int_<n> processes producing io on the disks slowing down the host leaving almost no room for organic io. If the NVMe is in, VMs tend to crash because of no more IO -timeleft for them., because the NVMe accelerates the problem a lot.
Some times the behavior stops after some hours (up to 24.) Quite often it never stops until reboot.

Interestingly, the processes shown in iotop are called "wr" but iotop show "disk reads", though atop shows "writes".

![image](https://cloud.githubusercontent.com/assets/7711872/23458327/257e9eaa-fe7c-11e6-9563-fed41e58d8ac.png)

atop reports the disks to be very busy:

![image](https://cloud.githubusercontent.com/assets/7711872/23458627/4c83f440-fe7d-11e6-84d6-4f5a5599ce05.png)


I already posted in the Proxmox forum, but this seems to be a problem very specific on ZFS:

https://forum.proxmox.com/threads/unerkl%C3%A4rliches-phantom-write-i-o-mit-zfs.33284/

**Comment by gkkovacs** [comment]
We have been experiencing a similar issue on Proxmox 4.x since last year. We even opened threads about it, this is the latest: https://forum.proxmox.com/threads/kvm-guests-on-zfs-freeze-during-backup-restore.34362/

**We now think it's an IO scheduling and caching issue in ZFS.**
It looks like the following issues are about the same problem: #5867  #5857

**Reproduction**
Basically, if you have a few KVM guests' RAW disks on ZFS Zvols, and you start to restore another KVM guest to the same host (so heavy write IO is happening for extended periods), the running guests will get starved of IO for so long that their virtual memory subsystem will trip up and all kinds of problems start to appear: CPU soft locks, hung tasks, rcu_scheduler stalls, and all around slowed down disk and network IO (Windows guests regularly lose their RDP connections for several seconds). Extended heavy reads also cause some IO starvation, but it's the writes that really cause problems.

**Mitigation**
We have tried mitigating this by limiting host and guest swapping, increasing guest vm buffers, and tuning power management, but nothing solved it so far.

We have also recognized that **QCOW2 disks with cache=writeback are much less sensitive to the IO starvation** from the KVM guests' point of view than ZVOL RAW disks with cache=none, but it's not yet clear why:
- either ZFS schedules parallel file IO better between files compared to parallel IO between ZVOLs,
- or simply the writeback cache solves the guests problems by caching their swap efficiently to evade detection.
